### PR TITLE
refactor(stream,agg): add `StateTableColumnMapping` to convert column indices

### PR DESCRIPTION
**Pull request by richardchien**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR adds a new `StateTableColumnMapping` to convert column indices between state table and input chunk. It's intended to be used in managed agg states to find columns they're interested in from state table and construct state table rows from input chunks.

This is part of the refactoring of managed agg states.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

#4185

### CODE REVIEW

File: src/stream/src/common/column_mapping.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        Self {
+            upstream_columns,
+            mapping,
+        }
+    }
+
+    /// Convert state table column index to upstream chunk column index.
+    pub fn state_table_to_upstream(&self, idx: usize) -> Option<usize> {
+        self.upstream_columns.get(idx).copied()
+    }
```
Comment (st1page) [inline src/stream/src/common/column_mapping.rs:41]: does the function really be used? maybe we can remove it. If you want maintain it, I suggest to change it definition to return a size and panic when out of range

**Review by st1page — APPROVED** [review]
rest LGTM

**Review by BowenXiao1999 — APPROVED** [review]

File: src/stream/src/common/column_mapping.rs (hunk 1 above)
Comment (richardchien) [reply src/stream/src/common/column_mapping.rs:41]: It'll be used in later PR. I think it's better to maintain symmetricity between `state_table_to_upstream` and `upstream_to_state_table`? And also panic usually happens in `impl Index` but not struct methods?

File: src/stream/src/common/column_mapping.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    /// Convert upstream chunk column index to state table column index.
+    pub fn upstream_to_state_table(&self, idx: usize) -> Option<usize> {
+        self.mapping.get(&idx).copied()
+    }
+
+    /// Return the number of columns in the mapping.
+    pub fn size(&self) -> usize {
```
Comment (BugenZhao) [inline src/stream/src/common/column_mapping.rs:49]: Better to name it `len` here. 😄
```suggestion
    pub fn len(&self) -> usize {
```

File: src/stream/src/common/column_mapping.rs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+use std::collections::HashMap;
+
+pub struct StateTableColumnMapping {
+    /// index: state table column index, value: upstream column index
+    upstream_columns: Vec<usize>,
+    /// key: upstream column index, value: state table column index
+    mapping: HashMap<usize, usize>,
+}
+
+impl StateTableColumnMapping {
```
[the remark this answers is not in the document]
Comment (richardchien) [reply src/stream/src/common/column_mapping.rs:24]: it's already in `stream` 😇

[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/stream/src/common/column_mapping.rs:24]: My eyes are blind. 🤣🤣🤣🤣

**Review by BugenZhao — APPROVED** [review]

File: src/stream/src/common/column_mapping.rs (hunk 2 above)
Comment (richardchien) [reply src/stream/src/common/column_mapping.rs:49]: ok

**Pull request merged**